src/group_theory/subgroup.lean

```diff
@@ -1662,6 +1717,143 @@ set.subset.antisymm
     by { rw sup_eq_closure, apply Inf_le _, dsimp, refl })
   ((sup_eq_closure N H).symm ▸ subset_closure)
 
-end subgroup
+@[to_additive] lemma mem_sup_iff
+  {H K : subgroup G} {g : G} (h : ↑(H ⊔ K) = (H : set G) * K) :
+  g ∈ H ⊔ K ↔ ∃ x y, x ∈ H ∧ y ∈ K ∧ x * y = g :=
+set.ext_iff.1 h g
+
+@[to_additive] lemma mem_sup_iff' {H K : subgroup G} {g : G} (h : ↑(H ⊔ K) = (H : set G) * K) :
+  g ∈ H ⊔ K ↔ ∃ (x:H) (y:K), (x * y : G) = g :=
+(mem_sup_iff h).trans
+  ⟨λ ⟨a, b, ha, hb, h⟩, ⟨⟨a, ha⟩, ⟨b, hb⟩, h⟩,
+  λ ⟨⟨a, ha⟩, ⟨b, hb⟩, h⟩, ⟨a, b, ha, hb, h⟩⟩
+
+@[to_additive] lemma mod_law_left (A' A B : subgroup G) (hA : A' ≤ A) :
+  (A' : set G) * (A ⊓ B : subgroup G) = A ⊓ (A' * B) :=
+begin
+  simp only [coe_inf, set.inf_eq_inter],
+  rw set_like.le_def at hA,
+  ext,
+  refine ⟨λ h, set.mem_inter _ _, λ h, _⟩,
+  { rcases set.mem_mul.mp h with ⟨y, z, hy, hz, hyz⟩,
+    rw set.mem_inter_iff at hz,
+    have := mul_mem A (hA hy) hz.1,
+    rwa hyz at this, },
+  { rcases set.mem_mul.mp h with ⟨y, z, hy, hz, hyz⟩,
+    rw set.mem_mul,
+    exact ⟨y, z, hy, (set.mem_of_subset_of_mem (set.inter_subset_right ↑A ↑B) hz), hyz⟩, },
+  rw set.mem_inter_iff at h,
+  rcases h.2 with ⟨y, z, hy, hz, hyz⟩,
+  rw set.mem_mul,
+  refine ⟨x * z⁻¹, z, _, (set.mem_inter _ hz), _⟩,
+  rwa ← eq_mul_inv_of_mul_eq hyz,
+  have := eq_inv_mul_of_mul_eq hyz,
+  have boo := mul_mem A (inv_mem A (hA hy)) h.1,
+  rwa ← this at boo,
+  simp only [inv_mul_cancel_right],
+end
```

Comment:
A bit of proof golfing. The main trick here is to use that `rcases`/`rintros`/`obtain` can pattern match on `y * z = x` to replace `x` everywhere with `y * z`.
```suggestion
begin
  ext,
  simp only [coe_inf, set.inf_eq_inter, set.mem_mul, set.mem_inter_iff],
  split,
  { rintros ⟨y, z, hy, ⟨hzA, hzB⟩, rfl⟩,
    refine ⟨mul_mem A (hA hy) hzA, _⟩,
    exact ⟨y, z, hy, hzB, rfl⟩ },
  rintros ⟨hyz, y, z, hy, hz, rfl⟩,
  refine ⟨y, z, hy, ⟨_, hz⟩, rfl⟩,
  suffices : y⁻¹ * (y * z) ∈ A, { simpa },
  exact mul_mem A (inv_mem A (hA hy)) hyz
end
```